### PR TITLE
Implement support for Rainfall Runoff definition file (.fnm)

### PR DESCRIPTION
**Pull request by BeardedPlatypus**

### CODE REVIEW

File: hydrolib/core/io/fnm/models.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    controller_definitions: Path = Path("contr3b.def")
+    structure_tables: Path = Path("struct3b.tbl")
+    boundary_data: Path = Path("bound3b.3b")
+    boundary_tables: Path = Path("bound3b.tbl")
+    sobek_location_rtc: Path = Path("sbk_loc.rtc")  # TODO 58.
+    wwtp_data: Path = Path("wwtp.3b")
+    wwtp_tables: Path = Path("wwtp.tbl")
+    industry_general: Path = Path("industry.3b")
+
+    mappix_paved_area_sewage_storage_name: Literal["pvstordt.his"] = "pvstordt.his"
```
[the remark this answers is not in the document]
Comment (evetion) [reply hydrolib/core/io/fnm/models.py:81]: With some more reading, what about `Literal[Path("pvstordt.his")]`? 🤯

File: hydrolib/core/io/fnm/serializer.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        {}    * 119. Cel output file
+        {}    * 120. RR Log file for Simulate
+        {}    * 121. coupling WQ salt RTC
+        {}    * 122. RR Boundary conditions file for SOBEK3
+        {}    * 123. Optional RR ASCII restart (test) for OpenDA
+        {}    * 124. Optional LGSI cachefile
+        {}    * 125. Optional meteo NetCdf timeseries inputfile rainfall
+        {}    * 126. Optional meteo NetCdf timeseries inputfile evaporation
+        {}    * 127. Optional meteo NetCdf timeseries inputfile temperature (only for RR-HBV)
+        """.format(*values))
```
Comment (evetion) [inline hydrolib/core/io/fnm/serializer.py:179]: I don't like use of order, rather use the f string variant with keywords (can also be `format` though). And what happens on `None` for the optional fields? Should these be left empty `""` or will there be `None` printed?

File: hydrolib/core/io/fnm/parser.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+"""parser.py defines the read method for the RainfallRunoffModel."""
```
Comment (evetion) [inline hydrolib/core/io/fnm/parser.py:1]: Well this is by far the shortest parser you have ever written, nicely done!

**Review by evetion — CHANGES_REQUESTED** [review]

File: hydrolib/core/io/fnm/models.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    mappix_flow_rates_over_structures_name: Literal["strflodt.his"] = "strflodt.his"
+    mappix_flow_rates_to_edge_name: Literal["bndflodt.his"] = "bndflodt.his"
+    mappix_pluvius_max_sewage_storage_name: Literal["plvstrdt.his"] = "plvstrdt.his"
+    mappix_pluvius_max_flow_rates_name: Literal["plvflodt.his"] = "plvflodt.his"
+    mappix_balance_name: Literal["balansdt.his"] = "balansdt.his"
+    mappix_cumulative_balance_name: Literal["cumbaldt.his"] = "cumbaldt.his"
+    mappix_salt_concentrations_name: Literal["saltdt.his"] = "saltdt.his"
+
+    @property
+    def mappix_paved_area_sewage_storage(self) -> Path:
```
[the remark this answers is not in the document]
Comment (BeardedPlatypus) [reply hydrolib/core/io/fnm/models.py:100]: They are still paths, just paths that always have the same value. The values are still part of the model though (and the user should be informed if they are invalid or missing).

File: hydrolib/core/io/fnm/models.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    bui_file_for_continuous_calculation_series: Path = Path("NWRWCONT.#")
+    nwrw_output: Path = Path("NwrwSys.His")
+    rr_routing_link_definitions: Path = Path("3b_rout.3b")
+    cell_input_file: Path = Path("3b_cel.3b")
+    cell_output_file: Path = Path("3b_cel.his")
+    rr_simulate_log_file: Path = Path("sobek3b_progress.txt")
+    rtc_coupling_wq_salt: Path = Path("wqrtc.his")
+    rr_boundary_conditions_sobek3: Path = Path("BoundaryConditions.bc")
+
+    rr_ascii_restart_openda: Optional[Path] = None
```
[the remark this answers is not in the document]
Comment (BeardedPlatypus) [reply hydrolib/core/io/fnm/models.py:206]: According to the issue description #46 these are only optional ones.

File: hydrolib/core/io/fnm/serializer.py (hunk 2 above)
Comment (BeardedPlatypus) [reply hydrolib/core/io/fnm/serializer.py:179]: the `values` variable is processed see the line 
```python
values = (_get_formatted_value(v, max_len) for v in data.values())
```
which will handle None values as well as the empty space offset such that everything is nicely aligned.

f-strings could work as well, the main reason I opted for this was because the order is important for the files, and it would be harder to read what is parsed where, as all the lines would not be aligned anymore. But I agree that the implicit dependency on the order is not ideal.

Comment (BeardedPlatypus) [reply hydrolib/core/io/fnm/serializer.py:179]: Another problem is that we only pass the dictionary here, and not the model itself, thus all the variables would then be defined as `data["property-name"]` which I'm not super happy about either.

File: hydrolib/core/io/fnm/serializer.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    """Serialize the specified model.
+
+    Args:
+        model (RainfallRunoffModel): The model to serialize.
+
+    Returns:
+        str: The serialized RainfallRunoffModel in .fnm format.
+    """
+    if "filepath" in data:
+        del data["filepath"]
```
[the remark this answers is not in the document]
Comment (BeardedPlatypus) [reply hydrolib/core/io/fnm/serializer.py:33]: I wasn't entirely sure where to put this. My initial approach just had the model as an argument for the serializer. This gave me a bit more readable code as I would just use the fields and properties here. However, due to the current implementation of the (file) models, all serializer seem to work on their data dict. I was not sure where exactly this should have been filtered, and adding it here seemed to work correctly.

File: hydrolib/core/io/fnm/models.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    lgsi_cachefile: Optional[Path] = None
+    meteo_input_file_rainfall: Optional[Path] = None
+    meteo_input_file_evaporation: Optional[Path] = None
+    meteo_input_file_temperature: Optional[Path] = None
+
+    @classmethod
+    def property_keys(cls) -> Iterable[str]:
+        # Skip first element corresponding with file_path introduced by the FileModel.
+        return list(cls.schema()["properties"].keys())[1:]
```
[the remark this answers is not in the document]
Comment (BeardedPlatypus) [reply hydrolib/core/io/fnm/models.py:215]: Wouldn't pop still rely on the same ordering?

File: hydrolib/core/io/fnm/models.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    mappix_exceedance_time_reference_levels_name: Literal[
+        "ow_excdt.his"
+    ] = "ow_excdt.his"
+    mappix_flow_rates_over_structures_name: Literal["strflodt.his"] = "strflodt.his"
+    mappix_flow_rates_to_edge_name: Literal["bndflodt.his"] = "bndflodt.his"
+    mappix_pluvius_max_sewage_storage_name: Literal["plvstrdt.his"] = "plvstrdt.his"
+    mappix_pluvius_max_flow_rates_name: Literal["plvflodt.his"] = "plvflodt.his"
+    mappix_balance_name: Literal["balansdt.his"] = "balansdt.his"
+    mappix_cumulative_balance_name: Literal["cumbaldt.his"] = "cumbaldt.his"
+    mappix_salt_concentrations_name: Literal["saltdt.his"] = "saltdt.his"
```
Comment (BeardedPlatypus) [inline hydrolib/core/io/fnm/models.py:97]:
```suggestion
    # Mappix file names/paths should always be the same, as such they have been
    # defined as Literal values.
    mappix_paved_area_sewage_storage_name: Literal["pvstordt.his"] = "pvstordt.his"
    mappix_paved_area_flow_rates_name: Literal["pvflowdt.his"] = "pvflowdt.his"
    mappix_unpaved_area_flow_rates_name: Literal["upflowdt.his"] = "upflowdt.his"
    mappix_ground_water_levels_name: Literal["upgwlvdt.his"] = "upgwlvdt.his"
    mappix_green_house_bassins_storage_name: Literal["grnstrdt.his"] = "grnstrdt.his"
    mappix_green_house_bassins_results_name: Literal["grnflodt.his"] = "grnflodt.his"
    mappix_open_water_details_name: Literal["ow_lvldt.his"] = "ow_lvldt.his"
    mappix_exceedance_time_reference_levels_name: Literal[
        "ow_excdt.his"
    ] = "ow_excdt.his"
    mappix_flow_rates_over_structures_name: Literal["strflodt.his"] = "strflodt.his"
    mappix_flow_rates_to_edge_name: Literal["bndflodt.his"] = "bndflodt.his"
    mappix_pluvius_max_sewage_storage_name: Literal["plvstrdt.his"] = "plvstrdt.his"
    mappix_pluvius_max_flow_rates_name: Literal["plvflodt.his"] = "plvflodt.his"
    mappix_balance_name: Literal["balansdt.his"] = "balansdt.his"
    mappix_cumulative_balance_name: Literal["cumbaldt.his"] = "cumbaldt.his"
    mappix_salt_concentrations_name: Literal["saltdt.his"] = "saltdt.his"
```

File: hydrolib/core/io/fnm/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BeardedPlatypus) [reply hydrolib/core/io/fnm/models.py:84]: I have added the additional comment  to explain the reasoning for using Literals.

[the remark this answers is not in the document]
Comment (evetion) [reply hydrolib/core/io/fnm/models.py:84]: @BeardedPlatypus 

Because *all* these properties are Paths (right?), this can work.

```python
from pydantic import BaseModel
from pydantic import validator
from typing import Literal
from pathlib import Path



class A(BaseModel):
    a: Literal[Path("test.tif")]

    @validator("*", pre=True)
    def make_path(cls, v):
        if isinstance(v, str):
            return Path(v)
        return v


A(a=Path("test.tif"))  # works
A(a="test.tif")  # works
A(a="test.tif2")  # fails
```

**Review by evetion — APPROVED** [review]
